### PR TITLE
Fixed init-list construction when size_type is not int

### PR DESCRIPTION
**Pull request by tomalakgeretkal**
Rationale in code comments. I thought about adding a new unit test but it was a bit beyond me. Feel free to make suggestions!

Note that this has yet to be amalgamated (waiting for https://github.com/nlohmann/json/pull/4083).

### CODE REVIEW

File: include/nlohmann/json.hpp [hunk 1]
```diff
@@ -910,7 +910,10 @@ class basic_json // NOLINT(cppcoreguidelines-special-member-functions,hicpp-spec
         bool is_an_object = std::all_of(init.begin(), init.end(),
                                         [](const detail::json_ref<basic_json>& element_ref)
         {
-            return element_ref->is_array() && element_ref->size() == 2 && (*element_ref)[0].is_string();
+            // The cast is to ensure op[size_type] is called, bearing in mind size_type may not be int;
+            // (many string types can be constructed from 0 via its null-pointer guise, so we get a
+            // broken call to op[key_type], the wrong semantics and a 4804 warning on Windows)
+            return element_ref->is_array() && element_ref->size() == 2 && (*element_ref)[(size_type)0].is_string();
```
Comment (gregmarr) [inline include/nlohmann/json.hpp]: I'd recommend `size_type(0)` or `static_cast<size_type>(0)` rather than a c-style cast.

Comment (nlohmann) [reply include/nlohmann/json.hpp]: I agree - please change this.

Comment (tomalakgeretkal) [reply include/nlohmann/json.hpp]: Roger that.

Comment (tomalakgeretkal) [reply include/nlohmann/json.hpp]: 'tis done

**Comment by gregmarr** [comment]
#4083 has been merged

**Comment by tomalakgeretkal** [comment]
Amalgamated.

**Review by nlohmann — APPROVED** [review]
Looks good to me.

**Comment by nlohmann** [comment]
Thanks!

**Pull request merged**